pandas/core/indexing.py

```diff
@@ -1898,16 +1898,20 @@ def _setitem_with_indexer_2d_value(self, indexer, value):
 
         ilocs = self._ensure_iterable_column_indexer(indexer[1])
 
-        # GH#7551 Note that this coerces the dtype if we are mixed
-        value = np.array(value, dtype=object)
+        if not is_array_like(value):
+            # cast lists to array
+            value = np.array(value, dtype=object)
         if len(ilocs) != value.shape[1]:
             raise ValueError(
                 "Must have equal len keys and value when setting with an ndarray"
             )
 
         for i, loc in enumerate(ilocs):
-            # setting with a list, re-coerces
-            self._setitem_single_column(loc, value[:, i].tolist(), pi)
+            value_col = value[:, i]
+            if is_object_dtype(value_col):
+                # try coerce values as good as possible
```

Comment:
the wording here is pretty kludgy, maybe "casting to list so that we do type inference in setitem_single_column"

Reply:
Changed, thx